Client/Frontend/Settings/CustomSearchViewController.swift

```diff
@@ -0,0 +1,206 @@
+        spinnerView.snp_makeConstraints { make in
+            make.center.equalTo(view)
+        }
+        spinnerView.hidesWhenStopped = true
+    }
+
+    private func addSearchEngine(searchQuery: String, title: String) {
+        spinnerView.startAnimating()
+        createEngine(forQuery: searchQuery, andName: title).uponQueue(dispatch_get_main_queue()) { result in
+            self.spinnerView.stopAnimating()
+            guard let engine = result.successValue else {
+                let alert = ThirdPartySearchAlerts.incorrectCustomEngineForm()
+                self.presentViewController(alert, animated: true, completion: nil)
+                return
+            }
+            self.profile.searchEngines.addSearchEngine(engine)
+            self.navigationController?.popViewControllerAnimated(true)
+            SimpleToast().showAlertWithText(Strings.ThirdPartySearchEngineAdded)
+        }
+    }
+
+
+    func createEngine(forQuery query: String, andName name: String) -> Deferred<Maybe<OpenSearchEngine>> {
+        let deferred = Deferred<Maybe<OpenSearchEngine>>()
+        guard let template = getSearchTemplate(withString: query),
+            let url = NSURL(string: template.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLFragmentAllowedCharacterSet())!) where url.isWebPage() else {
+                deferred.fill(Maybe(failure: FaviconError()))
+                return deferred
+        }
+        FaviconFetcher.fetchFavImageForURL(forURL: url, profile: profile).uponQueue(dispatch_get_main_queue()) { result in
+            let image = result.successValue ?? FaviconFetcher.getDefaultFavicon(url)
+            let engine = OpenSearchEngine(engineID: nil, shortName: name, image: image, searchTemplate: template, suggestTemplate: nil, isCustomEngine: true)
+
+            //Make sure a valid scheme is used
+            let url = engine.searchURLForQuery("test")
+            let maybe = (url == nil) ? Maybe(failure: FaviconError()) : Maybe(success: engine)
+            deferred.fill(maybe)
+        }
+        return deferred
+    }
+
+    func getSearchTemplate(withString query: String) -> String? {
+        let SearchTermComponent = "%s"      //Placeholder in User Entered String
+        let placeholder = "{searchTerms}"   //Placeholder looked for when using Custom Search Engine in OpenSearch.swift
+
+        if query.contains(SearchTermComponent) {
+            return query.stringByReplacingOccurrencesOfString(SearchTermComponent, withString: placeholder)
+        }
+        return nil
+    }
+
+    override func generateSettings() -> [SettingSection] {
+
+        func URLFromString(string: String?) -> NSURL? {
+            guard let string = string else {
+                return nil
+            }
+            return NSURL(string: string)
+        }
+        
+        let titleField = CustomSearchEngineField(placeholder: "Title", settingDidChange: {fieldText in
+            self.engineTitle = fieldText!
```

Comment:
This makes me uneasy seeing us force unwrap an optional here. Can we `guard` + optional unwrap instead just in case?